Connectors/Autocad/Speckle.Connectors.Civil3dShared/HostApp/PropertySetBaker.cs

```diff
@@ -0,0 +1,369 @@
+  {
+    _settingsStore = settingsStore;
+    _logger = logger;
+    _propertyHandler = new PropertyHandler();
+  }
+
+  /// <summary>
+  /// Parse and bake all property set definitions from the root object.
+  /// Should be called once at the beginning of the receive operation.
+  /// </summary>
+  public void ParsePropertySetDefinitions(Base rootObject)
+  {
+    _propertySetDefinitionMap.Clear();
+
+    if (rootObject[ProxyKeys.PROPERTYSET_DEFINITIONS] is not Dictionary<string, object?> definitions)
+    {
+      return;
+    }
+
+    if (definitions.Count == 0)
+    {
+      return;
+    }
+
+    using var tr = _settingsStore.Current.Document.Database.TransactionManager.StartTransaction();
+
+    foreach (var definition in definitions)
+    {
+      string setName = definition.Key;
+      object? setDefObj = definition.Value;
+
+      if (setDefObj is not Dictionary<string, object?> setDefData)
+      {
+        _logger.LogWarning("Property set definition {SetName} has invalid data format", setName);
+        continue;
+      }
+
+      if (!setDefData.TryGetValue("propertyDefinitions", out var propDefsObj))
+      {
+        _logger.LogWarning("Property set definition {SetName} missing propertyDefinitions", setName);
+        continue;
+      }
+
+      if (propDefsObj is not Dictionary<string, object?> propertyDefinitions)
+      {
+        _logger.LogWarning("Property set definition {SetName} propertyDefinitions has invalid format", setName);
+        continue;
+      }
+
+      ADB.ObjectId defId = GetOrCreatePropertySetDefinition(setName, propertyDefinitions, tr);
+      if (!defId.IsNull)
+      {
+        _propertySetDefinitionMap[setName] = defId;
+      }
+    }
+
+    tr.Commit();
+  }
+
+  /// <summary>
+  /// Try to bake property sets from a Speckle object to a Civil3D entity.
+  /// </summary>
+  public bool TryBakePropertySets(ADB.Entity entity, Base sourceObject, ADB.Transaction tr)
+  {
+    if (
+      sourceObject["properties"] is not Dictionary<string, object?> properties
+      || !properties.TryGetValue("Property Sets", out var propertySetsObj)
+      || propertySetsObj is not Dictionary<string, object?> propertySets
+      || propertySets.Count == 0
+    )
+    {
+      return false;
+    }
+
+    try
+    {
+      foreach (var propertySet in propertySets)
+      {
+        string setName = propertySet.Key;
+        object? setDataObj = propertySet.Value;
+
+        if (setDataObj is not Dictionary<string, object?> setData)
+        {
+          _logger.LogWarning("Property set {SetName} has invalid data format", setName);
+          continue;
+        }
+
+        if (!TryBakePropertySet(entity, setName, setData, tr))
+        {
+          _logger.LogWarning("Failed to bake property set {SetName} onto entity", setName);
+        }
+      }
+
+      return true;
+    }
+    catch (Exception ex) when (!ex.IsFatal())
+    {
+      _logger.LogError(ex, "Failed to bake property sets onto entity {Handle}", entity.Handle);
```

Comment:
missing `$`

Reply:
My two cents on this -> [here](https://github.com/specklesystems/speckle-sharp-connectors/pull/1149#discussion_r2435446247)